targets.go

```diff
@@ -129,159 +167,319 @@ func ensureValueIndexable(t target) error {
-var mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+var (
+	sliceInterfaceType     = reflect.TypeOf([]interface{}{})
+	mapStringInterfaceType = reflect.TypeOf(map[string]interface{}{})
+)
 
-func ensureMapIfInterface(x target) {
+func ensureMapIfInterface(x target) error {
 	v := x.get()
+
 	if v.Kind() == reflect.Interface && v.IsNil() {
 		newElement := reflect.MakeMap(mapStringInterfaceType)
-		x.set(newElement)
+
+		err := x.set(newElement)
+		if err != nil {
+			return fmt.Errorf("ensureMapIfInterface: %w", err)
+		}
 	}
+
+	return nil
 }
 
+var errSetStringCanNotAssignString = errors.New("cannot assign string")
+
 func setString(t target, v string) error {
 	f := t.get()
 
 	switch f.Kind() {
 	case reflect.String:
-		return t.setString(v)
+		err := t.setString(v)
+		if err != nil {
+			return fmt.Errorf("setString: %w", err)
+		}
+
+		return nil
 	case reflect.Interface:
-		return t.set(reflect.ValueOf(v))
+		err := t.set(reflect.ValueOf(v))
+		if err != nil {
+			return fmt.Errorf("setString: %w", err)
+		}
+
+		return nil
 	default:
-		return fmt.Errorf("cannot assign string to a %s", f.Kind())
+		return fmt.Errorf("setString: %w to a %s", errSetStringCanNotAssignString, f.Kind())
 	}
 }
 
+var errSetBoolCanNotAssignBool = errors.New("cannot assign bool")
+
 func setBool(t target, v bool) error {
 	f := t.get()
 
 	switch f.Kind() {
 	case reflect.Bool:
-		return t.setBool(v)
+		err := t.setBool(v)
+		if err != nil {
+			return fmt.Errorf("setBool: %w", err)
+		}
+
+		return nil
 	case reflect.Interface:
-		return t.set(reflect.ValueOf(v))
+		err := t.set(reflect.ValueOf(v))
+		if err != nil {
+			return fmt.Errorf("setBool: %w", err)
+		}
+
+		return nil
 	default:
-		return fmt.Errorf("cannot assign bool to a %s", f.String())
+		return fmt.Errorf("setBool: %w to a %s", errSetBoolCanNotAssignBool, f.String())
 	}
 }
 
-const maxInt = int64(^uint(0) >> 1)
-const minInt = -maxInt - 1
+const (
+	maxInt = int64(^uint(0) >> 1)
+	minInt = -maxInt - 1
+)
 
+var (
+	errSetInt64InInt32     = errors.New("does not fit in an int32")
+	errSetInt64InInt16     = errors.New("does not fit in an int16")
+	errSetInt64InInt8      = errors.New("does not fit in an int8")
+	errSetInt64InInt       = errors.New("does not fit in an int")
+	errSetInt64InUint64    = errors.New("negative integer does not fit in an uint64")
+	errSetInt64InUint32    = errors.New("negative integer does not fit in an uint32")
+	errSetInt64InUint32Max = errors.New("integer does not fit in an uint32")
+	errSetInt64InUint16    = errors.New("negative integer does not fit in an uint16")
+	errSetInt64InUint16Max = errors.New("integer does not fit in an uint16")
+	errSetInt64InUint8     = errors.New("negative integer does not fit in an uint8")
+	errSetInt64InUint8Max  = errors.New("integer does not fit in an uint8")
+	errSetInt64InUint      = errors.New("negative integer does not fit in an uint")
+	errSetInt64Unknown     = errors.New("does not fit in an uint")
+)
```

Comment:
I didn't notice this until just now, but...if errors are only used in one place, I'd suggest they not be declared globally.  These global vars bloat package init costs for no real reason.

Reply:
I wish go had a better way of declaring errors...
unfortunately, if they are not placed outside of the function, then they really have no point to exist.
The goal is for other functions to be able to detect what errors was triggered. Some of these will be exported, but no way to know which ones right now..

Reply:
just did the test. The impact I see for 1000 declared errors is a 200kB in the generated binary.
What else are you concerned about that I should check @moorereason ?

Reply:
I'm talking about package init costs not binary size.  For example, given:

```go
package main

import (
        _ "github.com/pelletier/go-toml/v2"
)

func main() {
        return
}
```

Show package init costs with the Go toolchain on `master`:

```
$ git checkout master
$ GODEBUG=inittrace=1 go run . 2>&1 | grep '^init.*toml'
init github.com/pelletier/go-toml/v2 @0.79 ms, 0.015 ms clock, 288 bytes, 11 allocs
```
Now show init costs of this PR:
```
$ git checkout v2-golangci-lint
$ GODEBUG=inittrace=1 go run . 2>&1 | grep '^init.*toml'
init github.com/pelletier/go-toml/v2 @0.69 ms, 0.009 ms clock, 1736 bytes, 67 allocs
```
Those are the package init costs for simply importing `go-toml` into your project.  Go will run all init logic for all packages before it gets to running your code.  For Go libraries, we generally want that to be a small as possible.

Reply:
Makes complete sense!

So we can either:
* default errs outside functions and put them in if they don't need to be exported or used by other functions.
* default errs inside the functions and put them out when they need to be used/exported.

Followed by a last pass before v2 release to check which are used and which are exported.

I guess that's a @pelletier decision!

Reply:
One point of clarification:  I compared `master` to your PR, but I should have been comparing the `v2` branch...so:
```
GODEBUG=inittrace=1 go run . 2>&1 | grep '^init.*toml'
init github.com/pelletier/go-toml/v2 @1.4 ms, 0.022 ms clock, 1368 bytes, 44 allocs
```

Reply:
That's a good point. I haven't really considered the impact on initialization speed. As a rule, I favor keeping the "happy path" fast, even if it means being slower on the error path.

Let's remove them from the global init and declare them inline. We can also assume that they will not be exported. Ideally, most (all?) of them should really be `decodeError`s to provide the full context.